### PR TITLE
Add xWRF blogpost

### PR DESCRIPTION
**Pull request by lpilz**
This is our initial version for the xWRF blogpost. As this was originally planned as a myst notebook, we still have to include the rendered code or change its format.

### CODE REVIEW

**Comment by lpilz** [comment]
@jthielen Please feel free to amend this as you wish!

File: src/posts/xwrf/index.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+authors:
+  - name: Jon Thielen
+    github: jthielen
+  - name: Lukas Pilz
+    github: lpilz
+---
+
+![xWRF logo](https://github.com/xarray-contrib/xwrf/blob/main/docs/source/_static/xwrf_logo_bg_light.svg)
+
+In the past, when analyzing [Weather Research and Forecasting (`WRF`)](https://www.mmm.ucar.edu/weather-research-and-forecasting-model) data with python you would be restricted to using the [`wrf-python`](https://wrf-python.readthedocs.io/en/latest/) package. This is not very convenient, especially when working on very large datasets since it uses compiled Fortran code in the background essentially preventing the user from scaling to cluster-scale using utilities like [`dask`](https://www.dask.org/).
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:13]: I would refrain from being so critical of wrf-python... maybe reword as "while wrf-python has a rich set of functionality, it struggles when scaling up to larger datasets. It lacks direct integration with distributed/parallel tools, such as Dask, which enable the ability to work these larger terabyte or possibly even petabyte scale datasets."

File: src/posts/xwrf/index.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+In this post, we will show how `xWRF` works together with other utilities in order to post-process a large `WRF` output dataset.
+
++++
+
+## Installing `xWRF`
+
+Before we start using `xWRF`, we need to install it. We can do this by following these steps!
+
+1. Install this in your python environment using pip (`pip install xwrf`)
+2. Open up a notebook and use the imports shown below!
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:30]: It would be helpful to have an overview of which packages you are installing and why - ex. holoviews stack for visualization, xarray + xwrf for the io, dask for scaling up your analysis. pint_xarray and numpy for the calculations!

File: src/posts/xwrf/index.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import pint_xarray
+import numpy as np
+
+hv.extension('bokeh')
+```
+
+## Spin up a Cluster
+
+```{code-cell} ipython3
+cluster = SLURMCluster(name='dask-cluster',
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:49]: Can you also mention you can use a local cluster if desired? Or provide some background on which machine your are running this on?

File: src/posts/xwrf/index.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                      walltime='8:00:00',
+                      asynchronous=0)
+n_workers = 2
+cluster.scale(n_workers)
+client = Client(cluster)
+client.wait_for_workers(n_workers=n_workers)
+client
+```
+
+## Load the WRF output
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:64]: A brief mention of intake would be helpful here, as well as an overview of what this dataset is... why are you using it in this tutorial

File: src/posts/xwrf/index.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+```{code-cell} ipython3
+import intake
+cat = intake.open_catalog("https://raw.githubusercontent.com/xarray-contrib/xwrf-data/main/catalogs/catalog.yml")
+ssp5_ds = cat["xwrf-sample-ssp585"].to_dask()
+ssp5_ds
+```
+
+As one can see, this output is not very helpful. Heaps of necessary information like WRF model grid coordinates or coordinate index assignments are missing. But we can use `xWRF` to very efficiently fix this.
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:73]: Mention here that this is **just** using the regular `xr.open_dataset()`. This is currently abstracted from the user.

File: src/posts/xwrf/index.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+```{code-cell} ipython3
+%time
+ssp5_ds = ssp5_ds.xwrf.postprocess()
+ssp5_ds
+```
+
+As you can see, the `xWRF` postprocessing is done in no time at all thanks to it leveraging the `dask`'s integration and it's delayed computation feature. This is despite the uncompressed dataset being close to 60GB in size.
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:81]: I would move the explanation of the postprocessing up here

File: src/posts/xwrf/index.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+ssp5_ds.air_pressure
+```
+
+## Example analysis
+
++++
+
+Let's just imagine that we would want to investigate e.g. how climate change might affect the jet stream in different SSP scenarios. One comparison we might want to make could be between the SSP 585 and the SSP 245 simulations.
+
+So first, we load the SSP 245 dataset, too.
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:101]: An explanation of what these different SSP values mean would be helpful - even just using one of the figures from the IPCC report, and link to more info.

File: src/posts/xwrf/index.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    for field in ['U', 'V']:
+        ds[field] = ds[field].xwrf.destagger().variable
+    ds = ds.metpy.quantify()
+    _wind_speed = wind_speed(ds.U, ds.V).metpy.dequantify()
+    grid = xgcm.Grid(ds, periodic=False)
+    _wind_speed = grid.transform(_wind_speed, 'Z', plevs, target_data=ds.air_pressure.pint.to('hPa').metpy.dequantify(), method='log').persist()
+    wind_speeds_js.append(_wind_speed)
+```
+
++++ {"tags": []}
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:127]: Can this be moved elsewhere? Why here?

File: src/posts/xwrf/index.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    geo=True,
+    rasterize=True,
+    project=True
+)
+plot
+```
+
+![blogpost.gif](https://user-images.githubusercontent.com/14276158/191564410-1a7a83f0-3795-445a-9f5d-388ba225fa13.gif)
+
+Finally, we clean up our workspace.
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:154]: Can you add a conclusions statement here? Summarizing the state of the package, how it helped with this workflow, and how people can install + get involved.

File: src/posts/xwrf/index.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    github: jthielen
+  - name: Lukas Pilz
+    github: lpilz
+---
+
+![xWRF logo](https://github.com/xarray-contrib/xwrf/blob/main/docs/source/_static/xwrf_logo_bg_light.svg)
+
+In the past, when analyzing [Weather Research and Forecasting (`WRF`)](https://www.mmm.ucar.edu/weather-research-and-forecasting-model) data with python you would be restricted to using the [`wrf-python`](https://wrf-python.readthedocs.io/en/latest/) package. This is not very convenient, especially when working on very large datasets since it uses compiled Fortran code in the background essentially preventing the user from scaling to cluster-scale using utilities like [`dask`](https://www.dask.org/).
+
+Here, we present to you [`xWRF`](https://github.com/xarray-contrib/xwrf/), a lightweight interface for working with the model output in [Xarray](https://docs.xarray.dev/en/stable/). Using `xWRF` enables a seamless integration of the unique `WRF` data format into Xarray and the [Pangeo](https://pangeo.io/) software stack. It achieves this by:
```
Comment (mgrover1) [inline src/posts/xwrf/index.md:15]: Feel free to mention how far this has come since the prototype I wrote about last fall :)

**Review by mgrover1 — COMMENTED** [review]
Hey @lpilz - here are some comments!

File: src/posts/xwrf/index.md (hunk 1 above)
Comment (lpilz) [reply src/posts/xwrf/index.md]: Please check if you like it better this way.

File: src/posts/xwrf/index.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+In this post, we will show how `xWRF` works together with other utilities in order to post-process a large `WRF` output dataset.
+
++++
+
+## Installing `xWRF`
+
+Before we start using `xWRF`, we need to install it. We can do this by following these steps!
+
+1. Install this in your python environment using pip (`pip install xwrf`)
```
[the remark this answers is not in the document]
Comment (lpilz) [reply src/posts/xwrf/index.md]: Done!

File: src/posts/xwrf/index.md (hunk 2 above)
Comment (lpilz) [reply src/posts/xwrf/index.md:41]: Done!

File: src/posts/xwrf/index.md (hunk 3 above)
Comment (lpilz) [reply src/posts/xwrf/index.md]: I switched to a local cluster, since @jthielen wanted to run it this way. If it takes waay too long, I can also let it run on the HPC, I'd just need some hints as to how to get the rendered outputs into the blog post.

File: src/posts/xwrf/index.md (hunk 4 above)
Comment (lpilz) [reply src/posts/xwrf/index.md]: Added some info. Please check if the level of detail is okay. Also, I'm not sure if the analysis in and of itself is sensible (from a physical/GCM standpoint) - it was simply the first thing I could think of which showed off all the nice `xWRF` features.

File: src/posts/xwrf/index.md (hunk 5 above)
Comment (lpilz) [reply src/posts/xwrf/index.md]: Done! (Maybe I was a bit overprecise - we can iterate over this)

File: src/posts/xwrf/index.md (hunk 7 above)
Comment (lpilz) [reply src/posts/xwrf/index.md]: Added some more info

File: src/posts/xwrf/index.md (hunk 8 above)
Comment (lpilz) [reply src/posts/xwrf/index.md]: This is legacy from the `ipynb` export to `myst`. Removed it.

**Comment by lpilz** [comment]
I didn't manage to work in all improvements and will not be able to work on this until Tuesday, so please feel free to take over. If somebody needs write-access to my repo, I'll happily provide it.

Also @jthielen: the modifications I made to the code are completely untested, jsyk.

**Comment by jthielen** [comment]
I've updated this blog post based on @mgrover1's remaining suggestions as well as discussions @lpilz and I had separately. My changes also included embedding the HTML output from the demo's code (except for holoviews plot, which uses a separate GIF), which inflates the line count tremendously and may make the `.md` file difficult to review directly. Please do let me know if this is an appropriate approach for the xarray.dev blog, or if reduced text-based outupts would be better!

Also, @xarray-contrib/xwrf-devs, would you be able to take another review of this from the xWRF perspective? After that, we should be ready for review (moving this out of draft stage).

**Comment by lpilz** [comment]
@andersy005 @mgrover1 Could one of you give Vercel a poke, please?

**Comment by lpilz** [comment]
I don't know how to fix Vercel but I had a look at the draft and corrected some minor issues. I think it reads rather nicely :) Let's get it over the finish line, shall we?

@andersy005 @mgrover1 How can we access the Vercel logs to debug these issues?

**Comment by andersy005** [comment]
@lpilz, you should be able to get the logs by building the site locally. For this, you need to install yarn via conda for example 

```bash 
conda install -c conda-forge yarn
```


and run the following 


```bash 
yarn install
yarn dev

```

**Comment by lpilz** [comment]
I deployed it locally and got the error:
```
error - SyntaxError: unknown: Expected corresponding JSX closing tag for <table>. (119:0)

  117 | <td style="text-align: left;"><strong>Using processes:</strong> True</td>
  118 | `}</code></pre>
> 119 | </tr>
      | ^
  120 | <pre><code parentName="pre" {...{}}>{`    </tbody></table>
  121 |
  122 |     <details>
    at instantiate (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:72:32)
    at constructor (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:358:12)
    at Parser.raise (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:3335:19)
    at Parser.jsxParseElementAt (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:7984:16)
    at Parser.jsxParseElementAt (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:7946:32)
    at Parser.jsxParseElementAt (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:7946:32)
    at Parser.jsxParseElementAt (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:7946:32)
    at Parser.jsxParseElement (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:8015:17)
    at Parser.parseExprAtom (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:8029:19)
    at Parser.parseExprSubscripts (/home/lpilz/phd/work/xwrf/tutorials/xarray.dev/node_modules/@babel/parser/lib/index.js:12540:23) {
  code: 'BABEL_PARSE_ERROR',
  reasonCode: 'MissingClosingTagElement',
  syntaxPlugin: 'jsx',
  loc: Position { line: 119, column: 0, index: 10150 },
  pos: [Getter/Setter],
  page: '/blog/[id]'
}
```

@jthielen does that help?

**Comment by andersy005** [comment]
@jthielen / @lpilz, i'm curious... for the purpose of this post, is it necessary to include the Dask client's widget and the xarray html repr? would the text repr for xarray be enough?

**Comment by jthielen** [comment]
> @jthielen / @lpilz, i'm curious... for the purpose of this post, is it necessary to include the Dask client's widget and the xarray html repr? would the text repr for xarray be enough?

I thought they would be nice (especially if it meant more people could explore around what a WRF file usual contains), but yes, the text repr should be good enough (especially because the data variables are elided in the middle, rather than the end). I've updated the commits to include text rather than HTML rendered output, as well as a few quick updates based on what I saw in the local preview.

File: src/posts/introducing-xwrf/index.md [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+    clabel='wind speed [m/s]',
+    coastline=True,
+    geo=True,
+    rasterize=True,
+    project=True
+)
+plot
+```
+
+![Animation of wind speed difference over the western U.S. for various times and isobaric levels](/posts/introducing-xwrf/hvplot_output.gif)
```
Comment (andersy005) [inline src/posts/introducing-xwrf/index.md:366]: we should put the `gif` in the `public/posts/introductiong-xwrf` directory: https://github.com/xarray-contrib/xarray.dev/tree/main/public/posts

Comment (jthielen) [reply src/posts/introducing-xwrf/index.md:366]: Shouldn't it be there already as of https://github.com/xarray-contrib/xarray.dev/pull/343/commits/2b7e893aca76d9f4cd9922d3fc649b402482f951?

Comment (jthielen) [reply src/posts/introducing-xwrf/index.md:366]: Oh wait, never mind, messed up git...should be there now as of https://github.com/xarray-contrib/xarray.dev/pull/343/commits/0a5d4ae286c5d0b99a746c13494ab0d59758ffaf

**Comment by lpilz** [comment]
Great, I just had another look at the post and I think it's good to go :) We should amend the date, though, when we eventually publish...

Moving this out of draft mode!

**Comment by lpilz** [comment]
Sorry, didn't know that you couldn't request reviews from 2 people at the same time...

File: src/posts/introducing-xwrf/index.md [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+authors:
+  - name: Lukas Pilz
+    github: lpilz
+  - name: Jonathan (JT) Thielen
+    github: jthielen
+---
+
+![xWRF logo](https://raw.githubusercontent.com/xarray-contrib/xwrf/main/docs/source/_static/xwrf_logo_bg_light.svg)
+
+## TLDR
```
Comment (mgrover1) [inline src/posts/introducing-xwrf/index.md:13]: Can you put the full term here? For those who do not know TLDR?

**Review by mgrover1 — COMMENTED** [review]
One small suggestion!

File: src/posts/introducing-xwrf/index.md (hunk 13 above)
Comment (dcherian) [reply src/posts/introducing-xwrf/index.md:13]: SGTM ;)

Comment (mgrover1) [reply src/posts/introducing-xwrf/index.md:13]: Or summary/overview works :)

**Review by mgrover1 — APPROVED** [review]
Content looks great!!

**Review by andersy005 — APPROVED** [review]
Looks solid to me 👍🏽. Let me know if this is ready, and i'll do the honors...

**Comment by lpilz** [comment]
@jthielen If you're good, I'm ready for release! Thanks to everybody for your work :)

**Comment by dcherian** [comment]
pirate-pirate-ship

**Comment by jthielen** [comment]
> @jthielen If you're good, I'm ready for release! Thanks to everybody for your work :)

Yup, all good!
